### PR TITLE
out of body experience: avatar follows HMD position using velocity motor instead of micro teleports

### PR DESCRIPTION
**Pull request by AndrewMeadows**
The object of this PR is to make the avatar's physical body follow the HMD position using velocity, however it also has some cleanup and bug fixes for the CharacterGhostObject.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59e6ca8f8d0ae0dc15e155e35e0b918fde970d20 available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-59e6ca8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-59e6ca8.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7258835f3f481da9f698a03e2fbcf114adee8e3b available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-7258835.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-7258835.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ef8ef473404ecc9ac99d094cc651564570b83b2 available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-0ef8ef4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-0ef8ef4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 55e1d058876cf4bbfc9ab60f307b6ceb1d5fa8b4 available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-55e1d05.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-55e1d05.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3e02bac412d3d1ebe3b2bc3ccee3fbe9fc3da37f available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-3e02bac.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-3e02bac.dmg

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1229,6 +1229,8 @@ void MyAvatar::rebuildCollisionShape() {
     float scale = getUniformScale();
     float radius = scale * _skeletonModel->getBoundingCapsuleRadius();
     float height = scale * _skeletonModel->getBoundingCapsuleHeight() + 2.0f * radius;
+    const float CANONICAL_AVATAR_HEIGHT = 2.0f;
+    _canonicalScale = height / CANONICAL_AVATAR_HEIGHT;
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: This is not used yet, but my plan is to use this to scale tuned parameters with avatar scale.

File: libraries/physics/src/CharacterController.cpp [hunk 2]
```diff
@@ -202,68 +200,77 @@ const btScalar MIN_TARGET_SPEED_SQUARED = MIN_TARGET_SPEED * MIN_TARGET_SPEED;
...
+        float deltaAngle = acosf(glm::clamp(glm::dot(currentFacing, desiredFacing), -1.0f, 1.0f));
+        float angularSpeed = deltaAngle / FEW_SUBSTEPS;
+        if (angularSpeed > MAX_ANGULAR_SPEED) {
+            angularSpeed *= MAX_ANGULAR_SPEED / angularSpeed;
+        }
+        float sign = copysignf(1.0f, glm::dot(desiredFacing, currentRight));
+        btQuaternion angularDisplacement = btQuaternion(btVector3(0.0f, 1.0f, 0.0f), sign * angularSpeed * dt);
+        btQuaternion endRot = angularDisplacement * startRot;
+        _rigidBody->setWorldTransform(btTransform(endRot, startPos));
+    }
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterController.cpp]: This is where the avatar follow velocity is computed and is then applied as a "motor" for linear offsets.  The angular offset is still reduced using incremental teleports.

The velocity of the linear follow has four domains:

(1) Offsets larger than MAX_WALKING_SPEED_DISTANCE follow at MAX_WALKING_SPEED
(2) Between NORMAL_ and MAX_WALKING_SPEED_DISTANCE the velocity is interpolated between NORMAL and MAX.
(3) Below NORMAL_ it moves as NORMAL except...
(4) At very small distances it uses exponential decay (where it is very slow) for stability.

Note that the "timescales" of a "motor" can be tricky to think about.  It really represents the timescale that the motor will defeat whatever is the current velocity of the RigidBody (and therefore it also represents a relative strength against all other simultaneous motors, if any).

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3c5e13b34b48daf97b60291cbce8eece348d0784 available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-3c5e13b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-3c5e13b.dmg

File: libraries/physics/src/CharacterController.cpp [hunk 3]
```diff
@@ -576,131 +570,131 @@ void CharacterController::computeNewVelocity(btScalar dt, glm::vec3& velocity) {
...
-            // OUTOFBODY_HACK -- in collisionless state switch between Ground and Hover states
-            if (rayHasHit) {
-                SET_STATE(State::Ground, "collisionless above ground");
-            } else {
-                SET_STATE(State::Hover, "collisionless in air");
-            }
+            SET_STATE(State::Hover, "collisionless in air");
         }
     }
+}
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterController.cpp]: `updateState()` is new and contains a big chunk of logic that used to live in `preSimulation()`.

File: libraries/physics/src/CharacterGhostObject.cpp [hunk 4]
```diff
@@ -14,6 +14,8 @@
 #include <stdint.h>
 #include <assert.h>
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterGhostObject.cpp]: These changes below are some fixes to the Kinematic Character Controller logic -- mostly to remove the slight "jump up" at each step.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dfe828982b0963ac4260c2ee84cb7e86d680f3e6 available at: https://gist.github.com/56af7b838ece1de227f79c82a46db15f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-dfe8289.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8626-dfe8289.dmg

**Comment by hyperlogic** [comment]
👍

**Pull request merged**